---
title: "Containers and R: the Rockerverse and beyond"
author:
  - name: Daniel Nüst
    affiliation: Institute for Geoinformatics, University of Münster
    address:
    - Heisenbergstr. 2, 48149 Münster, Germany
    email: daniel.nuest@uni-muenster.de
  - name: Carl Boettiger
    #affiliation: -
    #email: -
  - name: Robrecht Cannoodt
    #affiliatin: -
    #email: -
  - name: Dirk Eddelbuettel
    #affiliation: -
    #email: -
  - name: Mark Edmondson
    #affiliation: -
    #email: -
  - name: Colin Fay
    #affiliation: -
    #email: -
  - name: Ben Marwick
    #affiliation: -
    #email: -
  - name: Karthik Ram
    #affiliation: -
    #email: -
  - name: Noam Ross
    #affiliation: -
    #email: -
  - name: Nan Xiao
    #affiliation: -
    #email: -
  - name: Lori Shepherd
    affiliation: Roswell Park Comprehensive Cancer Center
    address:
    - Elm & Carlton Streets, Buffalo, New York, 14263 USA
    email: lori.shepherd@roswellpark.org
  - name: Nitesh Turaga
    affiliation: Roswell Park Comprehensive Cancer Center
    address:
    - Elm & Carlton Streets, Buffalo, New York, 14263 USA
    email: nitesh.turaga@roswellpark.org
abstract: >
  The Rocker project provides widely-used Docker images for R across different application scenarios.
  This articles surveys describe downstream projects building upon Rocker.
  We also look beyond Rocker to other projects connecting containerisation with R.
  These use cases and the diversity of applications demonstrate the power of Rocker and containerisation for collaboration, effectivity, scalability, and transparency.
preamble: >
   \usepackage{longtable}
output: rticles::rjournal_article
---

## Introduction

The \citep{R} community keeps growing: the number of new packages on CRAN keeps on rising, meetups, conferences, online courses and unconferences prosper, and the uptake in education and industry increases [REFs].
All this cements `R`'s role as the lingua franca of statistics, data visualisation, and computational research.
Part of this success can be traced back to the _Rocker Project_ ([https://www.rocker-project.org/](https://www.rocker-project.org/)).
The community project was introduced in 2014 and provides a number of [Docker](https://en.wikipedia.org/wiki/Docker_(software)) images for various use cases \citet{RJ-2017-065}.
The considerable uptake and continued evolution of _Rocker_'s suite has lead to numerous projects extending or building upon _Rocker_ images ranging from reproducible research to production deployments.
This article presents this _Rockerverse_ of projects.
It also introduces related activities connecting the R language and environment with other containerisation solutions.
The main contributions is a coherent picture of the current lay of the land of containers in, with, and for `R`.
This diversity includes demonstrators, early prototypes, and mature projects.

## Containerization and Rocker

_Do we need a generic intro here?_

## Images for alternative R distributions [@nuest]

- [MRO](https://github.com/nuest/mro-docker)
- [Renjin](https://github.com/nuest/renjin-docker)
- [pqR](https://github.com/nuest/pqr-docker)
- [FastR](https://github.com/nuest/fastr-docker)

- [Bioconductor](http://bioconductor.org/help/docker/), https://github.com/Bioconductor/bioc_docker


#
# Draft bullet point notes (brainstorming session)
#

- Most system dependencies come from rocker/r-ver and rocker/rstudio
  as needed.

- Bioconductor works on two release cycles every year.

- Docker images based on rocker/rstudio for release and devel.

    - Naming and tagging of images.

- Community maintained docker images like proteomics and metabolomics
  and cytometry. (Include community names L gatto, sneumann, m.jiang)

- Linux nightly build replica (bioconductor_full).

    - testing environment for users and developers/maintainers.

    - Aim is to install MOST (if not all) Bioconductor packages.

    - Better alternative than distributing specific packages in a
      image.

    - These images will be used for cloud based solutions in the near
      future like AnVIL project.

    - distributed on singularity as well.

	- kubernetes orchestrated

- Improve reproducibility of Bioconductor versions and R versions
  which are closely tied together.

- Ruby and a rakefile along with a templating engine to generate all
  of our images once every 6 months. (link to src)

    - Takes away the pain of hand crafting dockerfiles each release cycl

    - recent work: improve src Dockerfiles, so that images are
      smaller. Improvements like cleaning cache etc. Not installing
      system dependencies once they have already been installed
      rocker - images.

- Docker hub allows us to build and distribute images to the
  community.

- Future:

    - Teaching workshops, and they don't have to worry about AWS
      working vs not.

    - Docker images used for distributing Bioconductor workshop
      material post conference.

#
# Rough draft - they asked one sentance per line
#


_Bioconductor_ is a free, open source and open development project for
the analysis and comprehension of genomic data [1]. The project
consists of 1741 R software packages as of August 15th 2019, as well
as packages containing annotation or experiment data. _Bioconductor_
has a semiannual release cycle, each release is associated with a
particular version of R. Docker images allow availability of current
and past versions of _Bioconductor_ for convenience and
reproducibility.

_Bioconductor_ docker images are built on top of rocker/r-ver and
rocker/rstudio. By using rocker as a base image, the R version and
most package and system dependencies are included in a _Bioconductor_
base image.  _Bioconductor_ installs packages based on the R version,
and therefore leverage the rocker/rstudio and rocker/r-ver use of R
version tagging. _Bioconductor_ selects the desired version of R from
rocker, adds the BiocManager CRAN package for installing appropriate
versions of _Bioconductor_ packages, and creates a _Bioconductor_
docker image with an informative tag (R\_version\_Bioc\_version). The
images are summarized on the _Bioconductor_ web site [5], maintained
in github [6], and available to the community through DockerHub [3].
All past and current combinations of R and _Bioconductor_ are
therefore accessible via a specific docker tag.

_Bioconductor_ has several other docker images specific to various
areas of research.  The 'core' image installs the most commonly used
_Bioconductor_ packages.  Other _Bioconductor_ images for proteomics,
metabolomics, and flow cytometry are community maintained.  All
community maintained images build on top of the _Bioconductor_ base
image and therefore indirectly the rocker images [4].

Bioconductor's _bioc_docker_ software works using a templating engine
based in ruby, has a source directory `src` and an output directory
`out`.[*] This takes away in large part the repeated handcrafting of
each Dockerfile produced by Bioconductor at each release cycle. Edits
to the Dockerfile need to be made in the source directory, based on a
config file, the output directory is populated with the appropriate
Dockerfile for each specific to the various areas of research. A
recent systematic "audit" of our docker images following best
practices (as issued by the Docker website) led to a reduction in the
number of layers and also size of the overall images being produced by
Bioconductor. Some of the optimizations include removing cache,
cleaning up after each layer produced by a `RUN` statement, and most
importantly, not re-installing any dependencies already provided by
rocker.

Bioconductor now produces a Docker image to mimic the _Bioconductor_
nightly build system on Linux. This image is called the
`bioconductor_full` (based on rocker/rstudio). The goal of the Docker
image is to install _all the system dependencies_ needed to install
all of the Bioconductor packages. We achieve this goal to a large
extent, and are able to greater than _99%_ of packages (1730 out of
1741). This image allows developers to test their packages extensively
by replicating errors on the nightly build system, and users to
install most of the packages they need without dealing with system
dependencies. There is the added bonus that it comes with an RStudio
interface to ease users into using R with Docker, thanks to rocker.

We found this to be a easier alternative to distributing installed
_Bioconductor_ packages within Docker images as this is not entirely
consistent with the way R/Bioconductor is used. Users usually install
and add packages to their scripts in real time. It also keeps the
Docker image smaller in size. Using R is almost never like a
'workflow', which is defined by the pre-installed packages. We
encourage users to mount a volume to their Bioconductor docker image,
where all the installed (compiled) packages are stored in the external
mountable volume. The library tree `.libPaths()` has
`/usr/local/lib/R/host-site-library` as the first location searched
for installed packages, which allows this functionality to work. This
enables the docker image to be updated by using a `docker pull`
command without the user having to worry about any packages saved on
the image.

The `bioconductor_full` image is forming the base for our cloud based
solutions in upcoming future projects. We are currently able to host a
Kubernetes orchestrated cluster on the Google Cloud Platform using the
bioconductor_full image and the rocker/r-ver images. The distribution
of computation is done through an R package called `RedisParam`
(github: mtmorgan/RedisParam), where the `bioconductor_full` image is
the master node and the `rocker/r-ver` images are the workers.

We envision these Docker images being the primary way of teaching in
the future. The goal is to let workshop participants pull a pre-built
docker image for the specific worksop and run them on their local
machines. As of right now, the same task is accomplished by creating
and spinning up multiple AWS images (one per participant). The docker
images can work in more fault tolerant, lasting, cost-effective
manner, and be used by participants throughout their learning
journey. The users can choose to scale their workshop skills using
their cloud accounts to scale up with this provided docker image
(using either the AWS Elastic Container service or Google Kubernetes
Engine etc.)

#
# Need to move these to bib file and properly format.
#

[1]  Gentleman, Robert C.; Carey, Vincent J.; Bates, Douglas M.; Bolstad, Ben;
Dettling, Marcel; Dudoit, Sandrine; Ellis, Byron; Gautier, Laurent; Ge,
Yongchao; Gentry, Jeff; Hornik, Kurt; Hothorn, Torsten; Huber, Wolfgang; Iacus,
Stefano; Irizarry, Rafael; Leisch, Friedrich; Li, Cheng; Maechler, Martin;
Rossini, Anthony J.; Sawitzki, Gunther; Smith, Colin; Smyth, Gordon; Tierney,
Luke; Yang, Jean Y. H.; Zhang, Jianhua (2004). "Bioconductor: open software
development for computational biology and bioinformatics". Genome Biology. 5
(10): R80. doi:10.1186/gb-2004-5-10-r80. PMC 545600

[2] R Core Team (2019). R: A language and environment for statistical
  computing. R Foundation for Statistical Computing, Vienna, Austria.
  URL https://www.R-project.org/

[3] ??DockerHub citation?

[4] ??Reference our docker help page/github?

[5] ?? https://bioconductor.org/help/docker/

[6] ?? https://github.com/Bioconductor/bioc_docker

[*] The original work on `bioc_docker` was done by Dan Tanenbaum


## Other containerisation platforms [NN]

- R images for Singularity
- Running Rocker images with
  - Singularity
  - CoreOS rkt?
- nix?

## Non-Debian images [NN]

- Alpine images
- Images used by R-Hub (overlap with CI?)
- [Windows](https://github.com/nuest/rocker-win) [@nuest]
- https://github.com/jlisic/R-docker-centos

## Rocker for Research Compendia [@benmarwick]

...

## Packages for automation and packaging [@nuest]

- [`containerit`](https://github.com/o2r-project/containerit/)
- [`dockertest`](https://github.com/traitecoevo/dockertest/)
- [`liftr`](https://liftr.me/) for OS-level reproducibility [@nanx]
- [`rize` for Shiny](https://github.com/cole-brokamp/rize)
- [`dockerfiler`](https://github.com/ColinFay/dockerfiler)

## Packages for control and provisioning [@nuest]

- [`stevedore`](https://github.com/richfitz/stevedore)
- [`babelwhale`](https://cran.r-project.org/web/packages/babelwhale/index.html): Running a Docker from R with Singularity or Docker as back-end. This is really useful in HPC environments where a user might not have root access but is able to install Singularity instead of Docker. [@rcannood]
- [`docker`](https://bhaskarvk.github.io/docker/)
- `RSelenium`
- [`googleComputeEngineR`](https://cloudyr.github.io/googleComputeEngineR/) (function `docker_run`)
- [`analogsea`](https://github.com/sckott/analogsea)
- [`harbor`](https://github.com/wch/harbor/)
- [`dockermachine`](https://github.com/cboettig/dockermachine)

## Continuous integration and continuous delivery [@noamross, @ColinFay]

- R-Hub
- DevOps
  - https://www.opencpu.org/posts/opencpu-with-docker/
- `r-ci`: https://github.com/ColinFay/r-ci
- [dynwrap](https://github.com/dynverse/dynwrap_containers/blob/master/.travis.yml) [@rcannood]
  - For this project, we use travis-ci to build rocker-derived containers, test them, and only push them to docker hub (from travis-ci.org) if the integration tests succeed.

## Common or public work environments [div]

- Binder
  - [`holepunch`](https://github.com/karthik/holepunch) streamlines making an R research compendium Binder-ready [@karthik]
  - [`repo2docker`](https://repo2docker.readthedocs.io/en/latest/config_files.html) (`install.R`, `DESCRPTION`, `runtime.txt`) based on the Rocker images [@nuest]
- GPU [@noamross, @cboettig]
- Gigantum stack
- in education (ready to use teaching/learning environments)
  - [`r-db`](https://github.com/ColinFay/r-db) [@ColinFay]
- [RCloud](https://github.com/att/rcloud/tree/master/docker) ?

## Deployment, processing, cloud [div]

- Docker images for cloud services [@MarkEdmondson1234]
- [`plumber`](https://www.rplumber.io/docs/hosting.html#docker)
- `batchtools` \citep{Lang2017batchtools} can [schedule jobs with Docker Swarm](https://mllg.github.io/batchtools/reference/makeClusterFunctionsDocker.html)
- scalable deployments, e.g. start with numerous Shiny talks mentioning Rocker at useR!2017
- [dynmethods](https://github.com/dynverse/dynmethods) [@rcannood]: In order to evaluate ±50 computational methods which all used different environments (R, Python, C++, ...), we wrapped each of them in a docker container and can execute these methods from R. Again, all of these containers are being built on travis-ci, and will only be pushed to docker hub if the integration test succeeds.
- [ShinyProxy](https://www.shinyproxy.io/) [creates a container](https://github.com/openanalytics/shinyproxy/blob/master/src/main/java/eu/openanalytics/services/DockerService.java#L388) for each user

## R versions [@ColinFay, @nuest]

- run code with Versions of R: https://srv.colinfay.me/r-online
- old R versions: https://github.com/rocker-org/rocker-versioned/issues/138
- [semantic version tags for Rocker](https://github.com/rocker-org/rocker-versioned/#version-tags)
- debugging memory problems across multiple versions: https://github.com/wch/r-debug
- cross-R testing: Show how you can run the same script in multiple R variants in containers (new contribution). Ideally the images should be created using `containerit` and controlled with `stevedore`/`docker`, i.e. via an R script.

## Discussion/outlook [all, please add bullet points]

- Missing pieces?
- consolidation (e.g. via packages using `dockerfiler` and `stevedore`)
- Common themes
  - reproducibility
- will knowledge about containers continue to spread?
- what is needed for even more containers with R?
- ...


\bibliography{RJreferences}
