---
title: "Containers and R: the Rockerverse and beyond"
author:
  - name: Daniel Nüst
    affiliation: Institute for Geoinformatics, University of Münster
    address:
    - Heisenbergstr. 2, 48149 Münster, Germany
    email: daniel.nuest@uni-muenster.de
  - name: Carl Boettiger
    #affiliation: -
    #email: -
  - name: Robrecht Cannoodt
    #affiliatin: -
    #email: -
  - name: Dirk Eddelbuettel
    #affiliation:
    #email: -
  - name: Mark Edmondson
    affiliation: IIH Nordic A/S, Google Developer Expert for GCP
    email: mark@markedmondson.me
  - name: Colin Fay
    affiliation: ThinkR
    email: contact@colinfay.me
  - name: Ben Marwick
    #affiliation: -
    #email: -
  - name: Karthik Ram
    #affiliation: -
    #email: -
  - name: Noam Ross
    #affiliation: -
    #email: -
  - name: Nan Xiao
    #affiliation: -
    #email: -
  - name: Lori Shepherd
    affiliation: Roswell Park Comprehensive Cancer Center
    address:
    - Elm & Carlton Streets, Buffalo, New York, 14263 USA
    email: lori.shepherd@roswellpark.org
  - name: Nitesh Turaga
    affiliation: Roswell Park Comprehensive Cancer Center
    address:
    - Elm & Carlton Streets, Buffalo, New York, 14263 USA
    email: nitesh.turaga@roswellpark.org
abstract: >
  The Rocker project provides widely-used Docker images for R across different application scenarios.
  This articles surveys describe downstream projects building upon Rocker.
  We also look beyond Rocker to other projects connecting containerisation with R.
  These use cases and the diversity of applications demonstrate the power of Rocker and containerisation for collaboration, effectivity, scalability, and transparency.
preamble: >
   \usepackage{longtable}
output: rticles::rjournal_article
---

## Introduction

The \citep{R} community keeps growing: the number of new packages on CRAN keeps on rising, meetups, conferences, online courses and unconferences prosper, and the uptake in education and industry increases [REFs].
All this cements `R`'s role as the lingua franca of statistics, data visualisation, and computational research.
Part of this success can be traced back to the _Rocker Project_ ([https://www.rocker-project.org/](https://www.rocker-project.org/)).
The community project was introduced in 2014 and provides a number of [Docker](https://en.wikipedia.org/wiki/Docker_(software)) images for various use cases \citet{RJ-2017-065}.
The considerable uptake and continued evolution of _Rocker_'s suite has lead to numerous projects extending or building upon _Rocker_ images ranging from reproducible research to production deployments.
This article presents this _Rockerverse_ of projects.
It also introduces related activities connecting the R language and environment with other containerisation solutions.
The main contributions is a coherent picture of the current lay of the land of containers in, with, and for `R`.
This diversity includes demonstrators, early prototypes, and mature projects.

## Containerization and Rocker

_Do we need a generic intro here?_

## Images for alternative R distributions [@nuest]

- [MRO](https://github.com/nuest/mro-docker)
- [Renjin](https://github.com/nuest/renjin-docker)
- [pqR](https://github.com/nuest/pqr-docker)
- [FastR](https://github.com/nuest/fastr-docker)

### Bioconductor

[_Bioconductor_](http://bioconductor.org/help/docker/) is an open source, open development project for the analysis and comprehension of genomic data [@gentleman_bioconductor_2004].
The project consists of 1741 R software packages as of August 15th 2019, as well as packages containing annotation or experiment data.
_Bioconductor_has a semiannual release cycle, each release is associated with a particular version of R.
Docker images allow availability of current and past versions of _Bioconductor_ for convenience and reproducibility.

_Bioconductor_ 'base' docker images are built on top of `rocker/r-ver` and `rocker/rstudio`.
_Bioconductor_ installs packages based on the R version, and therefore uses `rocker/rstudio` and `rocker/r-ver` version tagging.
_Bioconductor_ selects the desired version of R from _Rocker_, adds the BiocManager CRAN package for installing appropriate versions of _Bioconductor_ packages, and creates a _Bioconductor_ docker image with an informative tag (R\_version\_Bioc\_version).
The images are summarized on the _Bioconductor_ web site (https://bioconductor.org/help/docker/), maintained on GitHub ([https://github.com/Bioconductor/bioc_docker](https://github.com/Bioconductor/bioc_docker)), and available to the community through [DockerHub](https://hub.docker.com/).
Past and current combinations of R and _Bioconductor_ are therefore accessible via a specific docker tag.

_Bioconductor_ has several images in addition to 'base', specific to various areas of research.
The 'core' image installs the most commonly used _Bioconductor_ packages.
_Bioconductor_ images for proteomics, metabolomics, and flow cytometry are community maintained.
All community maintained images build on top of the _Bioconductor_ base image and therefore indirectly the _Rocker_ images.
To simplify building and maintaining _Bioconductor_ images, we use a Ruby templating engine.

A recent audit of our Dockerfiles, following best practices from the Docker website, led to a reduction in the size and number of layers in images we produce.
The most important insights involve the ['union' file system used by Docker](https://docs.docker.com/storage/storagedriver/overlayfs-driver/#how-container-reads-and-writes-work-with-overlay-or-overlay2).
In this file system, once a layer (e.g., `RUN` statement) writes to a file path, the file path is never altered.
A subsequent layer that might appear to remove or overwrite the path actually masks, rather than alters, the original.
It is therefore important to clean up (e.g., cache removal) within each layer, and to avoid re-installing existing dependencies.

A recent innovation is to produce a `bioconductor_full` image to emulate the _Bioconductor_ nightly Linux build machine.
The image contains the _system dependencies_ needed to install and check almost all (1730 of 1741) _Bioconductor_ software packages.
Users no longer have to manage complciated system dependencies.
The image is configured so that `.libPaths()` has `/usr/local/lib/R/host-site-library` as the first location.
Users mounting a location on the host file system to this location then persist installed packages across docker sessions or updates.
Many _R_ users pursue flexible work flows tailored to particular analysis needs, rather than standardized work flows.
The `bioconductor_full` image is well-suited to this pattern.
`bioconductor_full` provides developers with a test environment like _Bioconductor_'s build system.

Use of images suggests several interesting possibilities for the _Bioconductor_ project.
Images may be valuable in teaching, where participants pull pre-built images to avoid complicated configuration of their own computing environemnts.
An appeal of this over our current approach (providing Amazon Machine Instances for the duration of the course) is the utility of the image to participant after the course is over.
`bioconductor_full` introduces a common system configuration, so it becomes increasingly sensible for _Bioconductor_ to distribute convenient _binary_ packages.
Images also suggest approaches to more advanced computational models.
For instance, we are exploring use of images for [Helm](https://helm.sh/)-orchestrated [Kubernetes](https://kubernetes.io/) clusters on the Google Cloud Platform.
The user interacts with a manager image based on `bioconductor_full`, configured to perform map-reduce style computations via the BiocParallel package communicating with minimally-configured worker images.
A strength of this approach is that the responsibility for complex software configuration (including customized development) is shifted from the user to the experienced _Bioconductor_ core team.

## Other containerisation platforms [NN]

- R images for Singularity
- Running Rocker images with
  - Singularity
  - CoreOS rkt?
- nix?

## Non-Debian images [NN]

- Alpine images
- Images used by R-Hub (overlap with CI?)
- [Windows](https://github.com/nuest/rocker-win) [@nuest]
- https://github.com/jlisic/R-docker-centos

## Rocker for Research Compendia [@benmarwick]

...

## Packages for automation and packaging [@nuest]

- [`containerit`](https://github.com/o2r-project/containerit/)
- [`dockertest`](https://github.com/traitecoevo/dockertest/)
- [`liftr`](https://liftr.me/) for OS-level reproducibility [@nanx]
- [`rize` for Shiny](https://github.com/cole-brokamp/rize)
- [`dockerfiler`](https://github.com/ColinFay/dockerfiler)

## Packages for control and provisioning [@nuest]

- [`stevedore`](https://github.com/richfitz/stevedore)
- [`babelwhale`](https://cran.r-project.org/web/packages/babelwhale/index.html): Running a Docker from R with Singularity or Docker as back-end. This is really useful in HPC environments where a user might not have root access but is able to install Singularity instead of Docker. [@rcannood]
- [`docker`](https://bhaskarvk.github.io/docker/)
- `RSelenium`
- [`googleComputeEngineR`](https://cloudyr.github.io/googleComputeEngineR/) (function `docker_run`)
  - I don't think `docker_run` is main focus for containers for this package, its more how it uses Docker to launch custom environments that can be templated (e.g. rstudio, shiny etc.) and for parrallisation e.g. `library(future)`  using GCE VMs as endpoints. [@MarkEdmondson1234]
- [`analogsea`](https://github.com/sckott/analogsea)
- [`harbor`](https://github.com/wch/harbor/)
- [`dockermachine`](https://github.com/cboettig/dockermachine)

## Continuous integration and continuous delivery [@noamross, @ColinFay]

- R-Hub
- DevOps
  - https://www.opencpu.org/posts/opencpu-with-docker/
- `r-ci`: https://github.com/ColinFay/r-ci
- [dynwrap](https://github.com/dynverse/dynwrap_containers/blob/master/.travis.yml) [@rcannood]
  - For this project, we use travis-ci to build rocker-derived containers, test them, and only push them to docker hub (from travis-ci.org) if the integration tests succeed.
- Google Cloud Build [@MarkEdmondson1234]
  - https://cloud.google.com/cloud-build/
  - For me this is what makes Docker containers viable, as it builds the Dockerfiles on each GitHub commit.  It couples with Google Container Registry to build private Docker images, for downstream applications.

## Common or public work environments [div]

- Binder
  - [`holepunch`](https://github.com/karthik/holepunch) streamlines making an R research compendium Binder-ready [@karthik]
  - [`repo2docker`](https://repo2docker.readthedocs.io/en/latest/config_files.html) (`install.R`, `DESCRPTION`, `runtime.txt`) based on the Rocker images [@nuest]
- GPU [@noamross, @cboettig]
- Gigantum stack
- in education (ready to use teaching/learning environments)
  - [`r-db`](https://github.com/ColinFay/r-db) [@ColinFay]
- [RCloud](https://github.com/att/rcloud/tree/master/docker) ?

## Deployment, processing, cloud [div]

- Docker images for cloud services [@MarkEdmondson1234]
  - the most popular functionality for `googleComputeEngineR` is its use of Docker to enable parallel processing across VMS - [some demos here](https://cloudyr.github.io/googleComputeEngineR/articles/massive-parallel.html)
- Google Cloud Run - CaaS (Containers as a Service) that lets you launch a Docker container without worrying about underlying infrastructure. An R implementation is shown here at [cloudRunR](https://github.com/MarkEdmondson1234/cloudRunR) which uses it to create a scalable R plumber API. 
- [`plumber`](https://www.rplumber.io/docs/hosting.html#docker)
- `batchtools` \citep{Lang2017batchtools} can [schedule jobs with Docker Swarm](https://mllg.github.io/batchtools/reference/makeClusterFunctionsDocker.html)
- scalable deployments, e.g. start with numerous Shiny talks mentioning Rocker at useR!2017
- [dynmethods](https://github.com/dynverse/dynmethods) [@rcannood]: In order to evaluate ±50 computational methods which all used different environments (R, Python, C++, ...), we wrapped each of them in a docker container and can execute these methods from R. Again, all of these containers are being built on travis-ci, and will only be pushed to docker hub if the integration test succeeds.
- [ShinyProxy](https://www.shinyproxy.io/) [creates a container](https://github.com/openanalytics/shinyproxy/blob/master/src/main/java/eu/openanalytics/services/DockerService.java#L388) for each user

## R versions [@ColinFay, @nuest]

- run code with Versions of R: https://srv.colinfay.me/r-online
- old R versions: https://github.com/rocker-org/rocker-versioned/issues/138
- [semantic version tags for Rocker](https://github.com/rocker-org/rocker-versioned/#version-tags)
- debugging memory problems across multiple versions: https://github.com/wch/r-debug
- cross-R testing: Show how you can run the same script in multiple R variants in containers (new contribution). Ideally the images should be created using `containerit` and controlled with `stevedore`/`docker`, i.e. via an R script.

## Discussion/outlook [all, please add bullet points]

- Missing pieces?
- consolidation (e.g. via packages using `dockerfiler` and `stevedore`)
- Common themes
  - reproducibility
- will knowledge about containers continue to spread?
- what is needed for even more containers with R?
- the ability to move processing between services easily (e.g locally, one cloud providers VM, another cloud provider's Container-as-a-Service)
- ...


\bibliography{RJreferences}
